test/images/sample-device-plugin/sampledeviceplugin.go

```diff
@@ -94,17 +94,26 @@ func main() {
 
 	}
 	dp1.SetAllocFunc(stubAllocFunc)
+	var registerControlFile string
 
-	if registerControlFile := os.Getenv("REGISTER_CONTROL_FILE"); registerControlFile != "" {
+	if registerControlFile = os.Getenv("REGISTER_CONTROL_FILE"); registerControlFile != "" {
 		if err := handleRegistrationProcess(registerControlFile); err != nil {
 			panic(err)
 		}
+		if err := dp1.Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath); err != nil {
+			panic(err)
+		}
+		select {}
+	} else {
+		// if REGISTER_CONTROL_FILE environment variable is not explicitly specified, we would like to register the device
+		// plugin to kubelet as normal. The registration is added in a for loop to allow re-registration of the
+		// plugin to kubelet in case of node reboot or kubelet restarts.
+		for {
+			if err := dp1.Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath); err != nil {
+				panic(err)
+			}
+		}
```

Comment:
I fully agree with the intent of this PR. Got burned already by the plugin not reconnecting. Designwise full ack. Implementationwise, let's explore these two points:

1. I concur a registration busy loop is probably not the best approach, unless `Register` is blocking and the pluin can register just once (is this documented? I don't recall). Would be better to register, sleep, detect disconnection, attempt to reconnect until succeed
2. we need to clarify and make explicit if REGISTER_CONTROL_FILE controls _A_ registration or ALL registration attempts. IOW, should the plugin optionally block _every_ connection attempt including reconnections or just the first? All options are on the table, let's decide and document.

Reply:
> I fully agree with the intent of this PR. Got burned already by the plugin not reconnecting. Designwise full ack. Implementationwise, let's explore these two points:
> 
> 1. I concur a registration busy loop is probably not the best approach, unless `Register` is blocking and the pluin can register just once (is this documented? I don't recall). Would be better to register, sleep, detect disconnection, attempt to reconnect until succeed

Sure, wanted to have something quick and dirty to verify if the kubelet restart issue reported [here](https://github.com/kubernetes/kubernetes/issues/118559) is real or passed in the e2e tests due to the sample device plugin implementation. In addition to that, I wanted to get a general consensus on whether this is something we want to fix. Looks like we have that now, so let's totally improve the implementation.  

  > 2. we need to clarify and make explicit if REGISTER_CONTROL_FILE controls _A_ registration or ALL registration attempts. IOW, should the plugin optionally block _every_ connection attempt including reconnections or just the first? All options are on the table, let's decide and document.

With the introduction of the `REGISTER_CONTROL_FILE` the default behavior of automatic registration was changed. In the absence of control file or environment variable, we now don't register which might be unexpected and I was attempting to rectify that as part of this PR as well which might explain the rationale behind the title.

In my opinion we should have two flows:
1. Case 1: Where REGISTER_CONTROL_FILE is not specified or is empty: In this case we follow the normal flow of registering to kubelet by default. This is applicable for all the registrations of the plugin to kubelet.
2. Case 2: If the environment variable is specified with a control file, the deletion of the control file triggers registration. Again, this should happen for ALL the registrations when environment variable is specified.

Reply:
> Additionally, we need to review the expectations for device plugins. Are they expected to reconnect? [EDIT] [yes, they are.](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#handling-kubelet-restarts)[/EDIT]

Thanks for checking this.